drivers/wifi/siwx91x/siwx91x_wifi.c

```diff
@@ -709,6 +712,237 @@ static int siwx91x_stats(const struct device *dev, struct net_stats_wifi *stats)
 }
 #endif
 
+struct siwx91x_ps_profile {
+	/* HIGH_PERFORMANCE by default */
+	sl_si91x_performance_profile_t req_ps_mode;
+	uint16_t listen_interval;
+} siwx91x_ps_cfg;
+
+static int siwx91x_get_beacon_interval(const struct device *dev)
+{
+	sl_wifi_operational_statistics_t sl_stat = { };
+	struct siwx91x_dev *sidev = dev->data;
+	sl_status_t status;
+
+	if (sidev->state != WIFI_STATE_COMPLETED) {
+		LOG_ERR("Device not connected");
+		return -EIO;
+	}
+
+	status = sl_wifi_get_operational_statistics(SL_WIFI_CLIENT_INTERFACE, &sl_stat);
+	if (status != SL_STATUS_OK) {
+		LOG_ERR("Failed to get beacon interval: 0x%x", status);
+		return -EINVAL;
+	}
+
+	return sys_get_le16(sl_stat.beacon_interval);
+}
+
+static int siwx91x_set_power_save(const struct device *dev, struct wifi_ps_params *params)
+{
+	sl_wifi_performance_profile_t sl_ps_profile = { };
+	struct siwx91x_dev *sidev = dev->data;
+	sl_wifi_interface_t interface;
+	int beacon_interval;
+	sl_status_t status;
+
+	__ASSERT(params, "params cannot be NULL");
+
+	interface = sl_wifi_get_default_interface();
+	if (FIELD_GET(SIWX91X_INTERFACE_MASK, interface) != SL_WIFI_CLIENT_INTERFACE) {
+		params->fail_reason = WIFI_PS_PARAM_FAIL_OPERATION_NOT_SUPPORTED;
+		LOG_ERR("Wi-Fi not in station mode");
+		return -EINVAL;
+	}
+
+	if (sidev->state == WIFI_STATE_INTERFACE_DISABLED) {
+		params->fail_reason = WIFI_PS_PARAM_FAIL_CMD_EXEC_FAIL;
+		LOG_ERR("Command given in invalid state");
+		return -EINVAL;
+	}
+
+	get_wifi_current_performance_profile(&sl_ps_profile);
+
+	switch (params->type) {
+	case WIFI_PS_PARAM_STATE:
+		if (params->enabled == WIFI_PS_ENABLED) {
+			if (sl_ps_profile.profile == HIGH_PERFORMANCE) {
+				if (siwx91x_ps_cfg.req_ps_mode == HIGH_PERFORMANCE) {
+					sl_ps_profile.profile = ASSOCIATED_POWER_SAVE_LOW_LATENCY;
+				} else {
+					sl_ps_profile.profile = siwx91x_ps_cfg.req_ps_mode;
+				}
+			} else {
+				return -EALREADY;
+			}
+
+			if (sl_ps_profile.dtim_aligned_type == 0) {
+				beacon_interval = siwx91x_get_beacon_interval(dev);
+				if (beacon_interval < 0) {
+					params->fail_reason = WIFI_PS_PARAM_FAIL_DEVICE_NOT_CONNECTED;
```

Comment:
The compliance check is failing because of the long (> 100) lines. To me it looks like this function could use some refactoring into smaller helper functions, so that you don't need to do so deep indentations.